Only find aws hosts that are running

jsravn opened the pull request:
When finding instance by node name in AWS, only retrieve running
instances when using tags to identify nodes.  Otherwise terminated, old
nodes can show up with the same tag when rebuilding nodes in the
cluster.

This was extracted from #20207

balooo commented:
I can confirm that I authored these commits.

justinsb commented:
OK thanks for extracting - this makes it much clearer.

Two things we should figure out:

1. We have a function isAlive that we should probably remove and merge into findInstanceByNodeName.  Looking through the code, we don't care about retrieving non-alive instances.  This will make the code a little more efficient also!
1. I think that getInstancesByNodeNames should behave the same way as the singular form (although if this proves difficult, we can remove getInstancesByNodeNames in favor of getAllInstances; it's only used in the ELB code when we're actually getting almost all instances)

chbatey commented:
We've changed getInstancesByNodeNames to filter via the AWS API and rebased

justinsb commented:
Marking for 1.2 per request of jsravn in slack; not guaranteed to make 1.2 but has been highlighted as a serious issue.

justinsb commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -2274,35 +2250,35 @@ func (a *AWSCloud) getInstancesByIDs(instanceIDs []*string) (map[string]*ec2.Ins
 // This is currently implemented by fetching all the instances, because this is currently called for all nodes (i.e. the majority)
```
Can we update this comment please to reflect the more efficient implementation?

justinsb commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -2274,35 +2250,35 @@ func (a *AWSCloud) getInstancesByIDs(instanceIDs []*string) (map[string]*ec2.Ins
 // This is currently implemented by fetching all the instances, because this is currently called for all nodes (i.e. the majority)
 // In practice, the breakeven vs looping through and calling getInstanceByNodeName is probably around N=2.
 func (a *AWSCloud) getInstancesByNodeNames(nodeNames []string) ([]*ec2.Instance, error) {
-	allInstances, err := a.getAllInstances()
```
Is getAllInstances now unused?  If so, we can delete it.

justinsb commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -2274,35 +2250,35 @@ func (a *AWSCloud) getInstancesByIDs(instanceIDs []*string) (map[string]*ec2.Ins
-	allInstances, err := a.getAllInstances()
-	if err != nil {
-		return nil, err
+	names := make([]*string, len(nodeNames))
```
I think aws.StringSlice does this for you

justinsb commented:
A few nits, and a rebase needed (totally my fault - I solved one small instance of this problem & broke your more complete fix).  But then I will LGTM.  We'll also need the CLA-bot to be happy.

Commit:
Only find running aws hosts by nodename

When finding instance by node name in AWS, only retrieve running
instances.  Otherwise terminated, old nodes can show up with the same
tag when rebuilding nodes in the cluster.

Another improvement made is to filter instances by the node names
provided, rather than selecting all instances and filtering in code.

Authors: @jsravn, @chbatey, @balooo

justinsb commented:
LGTM - thanks!
